src/io/sqlite_store.rs

```diff
@@ -0,0 +1,228 @@
+		let sql =
+			format!("SELECT value FROM {} WHERE namespace=:namespace AND key=:key;", KV_TABLE_NAME);
+
+		let msg = format!("Failed to read from key: {}/{}", namespace, key);
+		let res = self
+			.connection
+			.lock()
+			.unwrap()
+			.query_row(
+				&sql,
+				named_params! {
+					":namespace": namespace,
+					":key": key,
+				},
+				|row| row.get(0),
+			)
+			.map_err(|_| std::io::Error::new(std::io::ErrorKind::NotFound, msg))?;
+		Ok(Cursor::new(res))
+	}
+
+	fn write(&self, namespace: &str, key: &str, buf: &[u8]) -> std::io::Result<()> {
+		let mut locked_conn = self.connection.lock().unwrap();
+
+		let msg = format!("Failed to start transaction");
+		let sql_tx = locked_conn
+			.transaction()
+			.map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?;
+
+		let sql = format!(
+			"INSERT OR REPLACE INTO {} (namespace, key, value) VALUES (:namespace, :key, :data);",
+			KV_TABLE_NAME
+		);
+		let msg = format!("Failed to write to key: {}/{}", namespace, key);
+		sql_tx
+			.execute(
+				&sql,
+				named_params! {
+					":namespace": namespace,
+					":key": key,
+					":data": buf,
+				},
+			)
+			.map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?;
+
+		let msg = format!("Failed to commit transaction");
+		sql_tx.commit().map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?;
+
+		Ok(())
+	}
+
+	fn remove(&self, namespace: &str, key: &str) -> std::io::Result<bool> {
+		let sql = format!("DELETE FROM {} WHERE namespace=:namespace AND key=:key;", KV_TABLE_NAME);
+		let msg = format!("Failed to delete key: {}/{}", namespace, key);
+		let changes = self
+			.connection
+			.lock()
+			.unwrap()
+			.execute(
+				&sql,
+				named_params! {
+					":namespace": namespace,
+					":key": key,
+				},
+			)
+			.map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?;
+
+		let was_present = changes != 0;
+
+		Ok(was_present)
+	}
+
+	fn list(&self, namespace: &str) -> std::io::Result<Vec<String>> {
+		let locked_conn = self.connection.lock().unwrap();
+
+		let sql = format!("SELECT key FROM {} WHERE namespace=:namespace", KV_TABLE_NAME);
+		let msg = format!("Failed to prepare statement");
+		let mut stmt = locked_conn
+			.prepare(&sql)
+			.map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?;
+
+		let mut keys = Vec::new();
+
+		let msg = format!("Failed to retrieve queried rows");
+		let rows_iter = stmt
+			.query_map(named_params! {":namespace": namespace, }, |row| row.get(0))
+			.map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?;
+
+		for k in rows_iter {
+			let msg = format!("Failed to retrieve queried rows");
+			keys.push(k.map_err(|_| std::io::Error::new(std::io::ErrorKind::Other, msg))?);
+		}
+
+		Ok(keys)
+	}
+}
+
+impl KVStorePersister for SqliteStore {
+	fn persist<W: Writeable>(&self, prefixed_key: &str, object: &W) -> lightning::io::Result<()> {
+		let msg = format!("Could not persist file for key {}.", prefixed_key);
```

Comment:
"having the different backends decide to rename namespaces is a bad idea". yes exactly
that is why we should not be using anything to do with parent directories and key parsing from ldk-key as namespace keys. (for all kvstores including filesystem)

afaiu, ldk doesn't enforce anything by having keys like "monitors/[txid][index]", ldk is just providing a prefix.
 we could have something like "ldk_core" as namespace and "monitors/[txid][index]" as full key. 
 
 otherwise we are letting ldk-core create namespaces dynamically instead of ldk-node having fixed set of namespaces.(even though we own ldk-core, this is not ideal from ldk-node user perspective)
 
otherwise, in future, if ldk decides to store something new as "xyzMeta/" we would end up creating a namespace out of it and migration code will need to change.

Reply:
>No, it's clearly not just a prefix but an implied namespace, actually even an implied path as all assumptions are made around FilesystemPersister

there is no enforced implied namespace, cash-lit doesn't care about it, ldk doesn't care about it. We are making that assumption right now in this PR, as we split the key into a file-path for a non-filesystem-directory datastore. We shouldn't be thinking of this having any relation to file-path or filesystem-directory/subdirectory.